packages/compass-web/polyfills/fs/index.ts

```diff
@@ -0,0 +1,12 @@
+export const promises = {
+  chmod() {
+    return Promise.resolve();
+  },
+};
+export function rmSync() {
```

Comment:
Just curious: Why do we need these at all? Surely no code path on the web should ever encounter this? Or is webpack just not smart enough to never include that?

Reply:
I'm sorry in advance for the long read, I thought I'd use the opportunity to provide as much context here as possible 😄

So there are a couple of things that affect what exactly we need to polyfill and how. From webpack perspective as soon as it sees an `import` or `require` statement in the code, it will try to resolve the module and pull the code in the bundle, even if it's not used*. And if the module is missing (like it happens with Node.js built-in ones, they are not "real" modules and won't work anyway), webpack will throw an error. This is what we are trying to solve by introducing this [special `/provider` export](https://docs.google.com/document/d/1wkurN8xz5P98jTf8ako_7tQqJG3dxeRpK8JA3Wru7ko/edit#bookmark=id.79i8aflu7ccy) for all things that are platform specific, with providers having no direct imports for platform specific stuff, only accessor functions (hooks and "service locators").

The thing is, this process of converting services in still in progress and we wanted to see how far we are and whether or not compass-web will work as expected, so we started setting it up a bit early and still have to provide "safe" aliases and polyfills for some platform specific imports.

Now there is also different ways of polyfilling something in webpack, you can provide an actual polyfill file to use in place of the original import, and there is also an option to set `<package name>: false` instead of a polyfill path and in that case webpack will just return an empty module (literally a `{}` instead of import). Then in Compass with platform-specific stuff we're running into two possible cases.

- Something is imported and unsed conditionally (like using an ssh connection)
- Something is imported and is used without condition, but doesn't affect the behavior if fails or does nothing on call (list mongodb-log-writer calling an `rmSync` method, or oidc connection trying to create http server even when the server connection being established is not of type oidc)

For the first case we use `alias: false`, for the second we are providing a bare minimum functioning polyfill. This particular `rmSync` is called by `mongodb-log-writer package`, but this should go away after we are done with [COMPASS-7407](https://jira.mongodb.org/browse/COMPASS-7407) (the service provider is already set up, but some places are still using direct imports of the package instead of a provider).

Also I think worh noting that in case of data-service behavior we didn't really thought through that our devtools-connect actually needs to be packaged for browser, and so this is why I'm planning to open tickets for that. We can do with polyfills for now, but I would prefer if we could at least do emty module polyfilling of some of those packages instead of maintaining this weird code.

* - asterisk for tree shaking and all those things, but in reality in most cases code is written and pre-compiled by typescript / babel at the point webpack parsing it in a way that it will pull everything in anyway and won't be able to tree shake it.

Reply:
Thanks! This is very useful.